sdks/how-to/how-to-authenticate-a-user.md

```diff
@@ -0,0 +1,488 @@
+she is who she pretends to be. 
+
+## Getting the validation code
+The validation code is sent to the user by the iCure Message Gateway, to identify the user properly. As Daenaerys decided 
+to authenticate by email, she can now check her emails to get this code.
+
+:::info
+
+In a future version of Cockpit, you will be able to edit the email and SMS templating for the authentication process. 
+For now, these have all a default template. 
+
+:::
+
+Once Daenaerys got her validation code, she can come back to your app and continue the process. 
+
+### Completing the authentication process
+First, create a RSA KeyPair for Daenaerys, to let her protect her data :
+<!-- file://code-samples/how-to/authenticate-user/index.mts snippet:Generate an RSA Keypair for the user-->
+```typescript
+const keyPair = await anonymousApi.cryptoApi.RSA.generateKeyPair()
+const userPublicKey = ua2hex(await anonymousApi.cryptoApi.RSA.exportKey(keyPair.publicKey, 'spki'))
+const userPrivateKey = ua2hex(await anonymousApi.cryptoApi.RSA.exportKey(keyPair.privateKey, 'pkcs8'))
+```
+
+Once it's done, to complete the Daenaerys authentication, call the next service providing the previous 
+`AuthenticationProcess` the validation code Daenaerys received by email and the keypair you just created : 
+<!-- file://code-samples/how-to/authenticate-user/index.mts snippet:Complete authentication process-->
+```typescript
+const authenticationResult = await anonymousApi.authenticationApi.completeAuthentication(
+  authProcess!,
+  validationCode,
+  [userPrivateKey, userPublicKey],
+  () => undefined
+)
+
+const authenticatedApi = authenticationResult.medTechApi
+
+console.log(`Your new user id: ${authenticationResult.userId}`)
+console.log(`Database id where new user was created: ${authenticationResult.groupId}`)
+console.log(`Your initialised MedTechAPI: ***\${authenticatedApi}***`)
+console.log(`RSA keypair of your new user: ***\${authenticationResult.keyPair}***`)
+console.log(`Token created to authenticate your new user: ***\${authenticationResult.token}***`)
+```
+
+As a result, you receive : 
+- The MedTechApi instance to use for Daenaerys (properly initialised);
+- The `userId`, identifying Daenaerys user uniquely;
+- The `groupId`, identifying the database in which Daenaerys was created;
+- The `keyPair`, the RSA keypair you created for Daenaerys; 
+- The `token`, the time-limited token created for Daenaerys, to authenticate her; 
+
+Make sure to save these elements to be able to authenticate Daenaerys again when she'll come back on your app.
+
+<!-- file://code-samples/how-to/authenticate-user/index.mts snippet:Save credentials-->
+```typescript
+// saveSecurely does not exist : Use your own way of storing the following data securely
+// One option is to put these elements into the localStorage
+saveSecurely(userEmail, authenticationResult.token, authenticationResult.userId, authenticationResult.groupId, authenticationResult.keyPair)
+```
+
```

Comment:
I think it would be a good idea to add here that if the private key is stored only in the browser local storage and if the user deletes is local browser data they may lose access to data.
It would be a good idea to suggest our user to add request to their app to tell their user to store the private key somewhere in their pc.
```suggestion

:::danger

If you are building a web app and store the private key only in your user's browser local storage you should
consider that if the user deletes their browser data they will lose access to the data they created in iCure. 
After completing user registration it might be a good idea to ask your user to store their private key
in a safe place in their filesystem, possibly encrypting it with a password.
Make sure your users understand they should never share this file with anyone.

:::
```